# Semi-Structure Data Analytics 

# Installation of the required library

In [15]:
%pip install pandas

Note: you may need to restart the kernel to use updated packages.


In [16]:
import pandas as pd 

## Import the csv file

In [17]:
df = pd.read_csv("sales.csv") 

In [19]:
# Show the table in sales 
df 

customer  \
0     {"gender":"M","age":42,"email":"cauho@witwuta....   
1     {"gender":"M","age":50,"email":"keecade@hem.uy...   
2     {"gender":"M","age":51,"email":"worbiduh@vowbu...   
3     {"gender":"F","age":45,"email":"vatires@ta.pe"...   
4     {"gender":"M","age":71,"email":"man@bob.mz","s...   
...                                                 ...   
4995  {"gender":"F","age":19,"email":"mohofri@cusa.j...   
4996  {"gender":"F","age":60,"email":"usur@eccu.cw",...   
4997  {"gender":"F","age":45,"email":"fili@vu.so","s...   
4998  {"gender":"F","age":47,"email":"dup@katjewtep....   
4999  {"gender":"M","age":33,"email":"lihzodud@voble...   

                                                  items purchaseMethod  \
0     [{"name":"printer paper","tags":["office","sta...         Online   
1     [{"name":"envelopes","tags":["stationary","off...          Phone   
2     [{"name":"notepad","tags":["office","writing",...       In store   
3     [{"name":"binder","tags":["school","general","...       In store   
4     [{"name":"envelopes","tags":["stationary","off...         Online   
...                                                 ...            ...   
4995  [{"name":"binder","tags":["school","general","...       In store   
4996  [{"name":"binder","tags":["school","general","...       In store   
4997  [{"name":"backpack","tags":["school","travel",...         Online   
4998  [{"name":"pens","tags":["writing","office","sc...       In store   
4999  [{"name":"envelopes","tags":["stationary","off...         Online   

                      saleDate storeLocation  
0     2015-03-23T21:06:49.506Z        Denver  
1     2015-08-25T10:01:02.918Z       Seattle  
2     2017-06-22T09:54:14.185Z        Denver  
3     2015-02-23T09:53:59.343Z       Seattle  
4     2014-03-31T16:02:06.624Z        Austin  
...                        ...           ...  
4995  2017-07-12T06:54:17.918Z        Denver  
4996  2017-11-10T00:17:46.358Z        Denver  
4997  2013-03-20T18:17:19.463Z        Denver  
4998  2014-11-29T10:09:41.573Z        Denver  
4999  2014-08-18T06:25:49.739Z      New York  

[5000 rows x 5 columns]

In [20]:
# Convert the string representations of dictionaries in the 'items' column to actual dictionaries
df['items'] = df['items'].apply(lambda x: eval(x) if isinstance(x, str) else x)

# Extract the 'name' field from the dictionaries
df['name'] = df['items'].apply(lambda x: x[0]['name'] if x and isinstance(x, list) and x[0] and 'name' in x[0] else None)

# Display the 'name' column
print(df['name'])

0       printer paper
1           envelopes
2             notepad
3              binder
4           envelopes
            ...      
4995           binder
4996           binder
4997         backpack
4998             pens
4999        envelopes
Name: name, Length: 5000, dtype: object


## Name from items row has been extracted

In [21]:
df['quantity'] = df['items'].apply(lambda x: x[0]['quantity'] if x else None)

# Now, the "name" column contains the "name" field from the "items" column
print(df['quantity'])

0        2
1       10
2        1
3        1
4        8
        ..
4995     5
4996     3
4997     1
4998     1
4999     5
Name: quantity, Length: 5000, dtype: int64


## Quantity from items row has been extracted

In [22]:
df['price'] = df['items'].apply(lambda x: x[0]['price'] if x else None)

# Now, the "name" column contains the "name" field from the "items" column
print(df['price'])

0        {'$numberDecimal': '40.01'}
1         {'$numberDecimal': '8.05'}
2        {'$numberDecimal': '30.08'}
3        {'$numberDecimal': '20.08'}
4        {'$numberDecimal': '21.95'}
                    ...             
4995     {'$numberDecimal': '23.38'}
4996     {'$numberDecimal': '19.66'}
4997    {'$numberDecimal': '148.77'}
4998     {'$numberDecimal': '63.23'}
4999     {'$numberDecimal': '24.12'}
Name: price, Length: 5000, dtype: object


## Price from items has been extracted

## Download name to csv

In [13]:
name_df = df[['name']]

# Save the "name" DataFrame as a CSV file
name_df.to_csv("names.csv", index=False)

## Download quantity to csv

In [14]:
name_df = df[['quantity']]

# Save the "name" DataFrame as a CSV file
name_df.to_csv("quantity.csv", index=False)

## Download price to csv

In [15]:
name_df = df[['price']]

# Save the "name" DataFrame as a CSV file
name_df.to_csv("price.csv", index=False)

## Now that have filtered out all the neccessary rows, we can compile it manually in excel. 

## Import the new csv file 

In [23]:
df = pd.read_csv("final_sales.csv") 

In [24]:
# View the table 
df 

Gender  Satisfaction           Name  Quantity   Price purchaseMethod  \
0         M             4  printer paper         2   40.01         Online   
1         M             5      envelopes        10    8.05          Phone   
2         M             5        notepad         1   30.08       In store   
3         F             3         binder         1   20.08       In store   
4         M             3      envelopes         8   21.95         Online   
...     ...           ...            ...       ...     ...            ...   
4995      F             5         binder         5   23.38       In store   
4996      F             4         binder         3   19.66       In store   
4997      F             5       backpack         1  148.77         Online   
4998      F             3           pens         1   63.23       In store   
4999      M             3      envelopes         5   24.12         Online   

     storeLocation    saleDate  
0           Denver  2015-03-23  
1          Seattle  2015-08-25  
2           Denver  2017-06-22  
3          Seattle  2015-02-23  
4           Austin  2014-03-31  
...            ...         ...  
4995        Denver  2017-07-12  
4996        Denver  2017-11-10  
4997        Denver  2013-03-20  
4998        Denver  2014-11-29  
4999      New York  2014-08-18  

[5000 rows x 8 columns]

In [25]:
# check the columns 
df.columns


Index(['Gender', 'Satisfaction', 'Name', 'Quantity', 'Price', 'purchaseMethod',
       'storeLocation', 'saleDate'],
      dtype='object')

## Show top 10 products 

In [26]:
# Calculate sales for each product
df['Sales'] = df['Quantity'] * df['Price']

# Sort the DataFrame by sales in descending order and select the top 10 rows
top_10_products = df.sort_values(by='Sales', ascending=False).head(10)

# Display the top 10 products
top_10_products[['Name', 'Sales']]


Name    Sales
2185  laptop  7966.55
4734  laptop  7878.80
3810  laptop  7758.65
4286  laptop  7724.45
2857  laptop  7702.25
2247  laptop  7696.70
19    laptop  7658.45
2966  laptop  7628.30
4250  laptop  7600.45
4064  laptop  7590.35

## We can see that the top 10 products accross all the stores are all laptop. The highest being 7966.55

## Show top 3 products (name) sales by store (location). 

In [21]:
# Calculate sales for each product
df['Sales'] = df['Quantity'] * df['Price']

# Group the data by 'storeLocation' and find the top 3 products by sales in each group
top_3_products_by_location = df.groupby('storeLocation').apply(lambda x: x.nlargest(3, 'Sales'))

# Reset the index to get a clean DataFrame
top_3_products_by_location = top_3_products_by_location.reset_index(drop=True)

# Display only the 'storeLocation', 'Name', and 'Sales' columns
top_3_products_by_location[['storeLocation', 'Name', 'Sales']]


storeLocation    Name    Sales
0         Austin  laptop  7878.80
1         Austin  laptop  7696.70
2         Austin  laptop  7586.55
3         Denver  laptop  7724.45
4         Denver  laptop  7590.35
5         Denver  laptop  7438.35
6         London  laptop  7658.45
7         London  laptop  7484.95
8         London  laptop  7375.80
9       New York  laptop  7560.05
10      New York  laptop  6682.80
11      New York  laptop  5121.00
12     San Diego  laptop  7600.45
13     San Diego  laptop  6345.70
14     San Diego  laptop  6233.08
15       Seattle  laptop  7966.55
16       Seattle  laptop  7758.65
17       Seattle  laptop  7702.25

## The top 3 products sales from each stores are all laptop

## Show purchased method by gender table 

In [22]:
# Create a cross-tabulation table of purchase method by gender
purchase_method_by_gender = pd.crosstab(df['purchaseMethod'], df['Gender'])

# Display the table
purchase_method_by_gender


Gender             F     M
purchaseMethod            
In store        1430  1389
Online           813   772
Phone            284   312

## We can see that there are more females buying in store and online than male by 41 and 41, respectively. While males has 28 more on phone.

## Show monthly total sales 

In [23]:
# Convert the 'saleDate' column to a datetime data type 
df['saleDate'] = pd.to_datetime(df['saleDate'])

# Create a new column for the sale month
df['SaleMonth'] = df['saleDate'].dt.to_period('M')

# Group the data by 'SaleMonth' and calculate the total sales for each month
monthly_total_sales = df.groupby('SaleMonth')['Sales'].sum()

# Display the monthly total sales
monthly_total_sales


SaleMonth
2013-01    38914.17
2013-02    25219.56
2013-03    41838.18
2013-04    40043.06
2013-05    47731.07
2013-06    36333.93
2013-07    52443.75
2013-08    19253.67
2013-09    23790.29
2013-10    30326.39
2013-11    27214.12
2013-12    45513.79
2014-01    11468.01
2014-02    23559.79
2014-03    36166.46
2014-04    24156.16
2014-05    23413.36
2014-06    16086.22
2014-07    22296.59
2014-08    39825.22
2014-09    17623.55
2014-10    30630.84
2014-11    42837.20
2014-12    29620.30
2015-01    25012.91
2015-02    38968.22
2015-03    41215.83
2015-04    28115.30
2015-05    30647.34
2015-06    31084.00
2015-07    37152.88
2015-08    40475.74
2015-09    24594.27
2015-10    38681.11
2015-11    45833.83
2015-12    31536.19
2016-01    42087.46
2016-02    20334.07
2016-03    33559.23
2016-04    16875.95
2016-05    41870.18
2016-06    21054.51
2016-07    26349.46
2016-08    16247.45
2016-09    16141.10
2016-10    24153.67
2016-11    30293.12
2016-12    33809.05
2017-01    21393.47
2017-02   

## From the given output we can also determine the earliest and latest month recorded in the data

## Show rankings of each store (location). 

In [24]:
# Calculate sales for each store location
df['Sales'] = df['Quantity'] * df['Price']

# Group the data by 'storeLocation' and calculate the total sales for each location
store_sales = df.groupby('storeLocation')['Sales'].sum()

# Rank the store locations based on their total sales in descending order and start from 1
store_rankings = store_sales.rank(ascending=False, method='min').astype(int)

# Create a DataFrame that combines store rankings and total sales
store_rankings_df = pd.DataFrame({'StoreRank': store_rankings, 'TotalSales': store_sales})

# Sort the DataFrame by the 'TotalSales' column in descending order
store_rankings_df = store_rankings_df.sort_values(by='TotalSales', ascending=False)

# Reset the index to get a clean DataFrame
store_rankings_df = store_rankings_df.reset_index()

# Display the combined DataFrame
store_rankings_df


storeLocation  StoreRank  TotalSales
0        Denver          1   543903.18
1       Seattle          2   437337.40
2        London          3   298405.93
3        Austin          4   279867.12
4      New York          5   148527.47
5     San Diego          6   143187.42

## This ranking shows that Denver has the highest rank based on its total sales of 543903.18 while San Diego being the lowest rank at 143187.42